src/supervisor3.erl

```diff
@@ -1009,7 +1019,7 @@ do_restart_delay({RestartType, Delay}, Reason, Child, State) ->
             %% will fail to find the child in ?restarting state,
             %% hence ignore the restart, which will in turn
             %% cause it to enter a permanent zombie state.
-            State
+            replace_child(Child#child{pid = ?delayed_restart(TRef)}, State)
```

Comment:
why do we need a new tag ?
should `?restarting(...)` work?

we don't have to cancel the timer when a restarting child is deleted,
a stale send after is simply discarded.

Reply:
ah, for `simple_one_for_one` children, it may get added back.
but I'm not sure why would anyone use `simple_one_for_one` with a delayed restart, we should perhaps drop it.

Reply:
New tag is needed because according to the spec {restarting, } tuple contains pid and I put a reference here. So I wouldn't mix these cases.

Yes, the restarting child is discarded, *unless* a new child with the same id has been created while the old was restarting. It leads to the situation when supervisor3 continuously tries to start child with old spec and fails with *already_started* error.